From a79fa29c0000cde0c03866dbcf4cdf616c177eaf Mon Sep 17 00:00:00 2001
From: Daniel Stone <daniels@collabora.com>
Date: Tue, 19 May 2015 20:02:41 +0100
Subject: [PATCH 05/11] compositor-drm: Track cursor_plane with a drm_plane

Change the type of cursor_plane from a weston_plane (base tracking
structure) to a drm_plane (wrapper containing additional DRM-specific
details), and make it a dynamically-allocated pointer.

Using the standard drm_plane allows us to reuse code which already deals
with drm_planes, e.g. a common cleanup function.

v2: Add legacy cursor_planes to plane_list so they get cleaned up
    properly.

Signed-off-by: Daniel Stone <daniels@collabora.com>
Signed-off-by: Bharathraj Nagaraju <snbraj@codeaurora.org>
(cherry picked from commit ed59dc23796f786b971019a6437bbca681c0c50a)

Change-Id: I0574810f8a376e31a6e7c5aa7602bae26bbe3fef
---
 src/compositor-drm.c | 96 ++++++++++++++++++++++++++++++++++++++++++----------
 1 file changed, 78 insertions(+), 18 deletions(-)

diff --git a/src/compositor-drm.c b/src/compositor-drm.c
index 35d828f..7d22c0d 100644
--- a/src/compositor-drm.c
+++ b/src/compositor-drm.c
@@ -264,7 +264,7 @@ struct drm_output {
 
 	struct gbm_surface *surface;
 	struct gbm_bo *cursor_bo[2];
-	struct weston_plane cursor_plane;
+	struct drm_plane *cursor_plane;
 	struct weston_plane fb_plane;
 	struct weston_view *cursor_view;
 	int current_cursor;
@@ -1069,7 +1069,8 @@ drm_output_repaint(struct weston_output *output_base,
 
 	output->page_flip_pending = 1;
 
-	drm_output_set_cursor(output);
+	if (output->cursor_plane)
+		drm_output_set_cursor(output);
 
 	/*
 	 * Now, update all the sprite surfaces
@@ -1492,12 +1493,14 @@ static struct weston_plane *
 drm_output_prepare_cursor_view(struct drm_output *output,
 			       struct weston_view *ev)
 {
-	struct drm_backend *b =
-		(struct drm_backend *)output->base.compositor->backend;
+		struct drm_backend *b =
+			(struct drm_backend *)output->base.compositor->backend;
 	struct weston_buffer_viewport *viewport = &ev->surface->buffer_viewport;
 
 	if (b->gbm == NULL)
 		return NULL;
+	if (output->cursor_plane == NULL)
+		return NULL;
 	if (output->base.transform != WL_OUTPUT_TRANSFORM_NORMAL)
 		return NULL;
 	if (viewport->buffer.scale != output->base.current_scale)
@@ -1518,7 +1521,7 @@ drm_output_prepare_cursor_view(struct drm_output *output,
 
 	output->cursor_view = ev;
 
-	return &output->cursor_plane;
+	return &output->cursor_plane->base;
 }
 
 /**
@@ -1569,6 +1572,8 @@ drm_output_set_cursor(struct drm_output *output)
 	struct gbm_bo *bo;
 	int x, y;
 
+	assert(output->cursor_plane);
+
 	output->cursor_view = NULL;
 	if (ev == NULL) {
 		drmModeSetCursor(b->drm.fd, output->crtc_id, 0, 0, 0);
@@ -1578,9 +1583,9 @@ drm_output_set_cursor(struct drm_output *output)
 	buffer = ev->surface->buffer_ref.buffer;
 
 	if (buffer &&
-	    pixman_region32_not_empty(&output->cursor_plane.damage)) {
-		pixman_region32_fini(&output->cursor_plane.damage);
-		pixman_region32_init(&output->cursor_plane.damage);
+	    pixman_region32_not_empty(&output->cursor_plane->base.damage)) {
+		pixman_region32_fini(&output->cursor_plane->base.damage);
+		pixman_region32_init(&output->cursor_plane->base.damage);
 		output->current_cursor ^= 1;
 		bo = output->cursor_bo[output->current_cursor];
 
@@ -1595,14 +1600,15 @@ drm_output_set_cursor(struct drm_output *output)
 
 	x = (ev->geometry.x - output->base.x) * output->base.current_scale;
 	y = (ev->geometry.y - output->base.y) * output->base.current_scale;
-	if (output->cursor_plane.x != x || output->cursor_plane.y != y) {
+	if (output->cursor_plane->base.x != x ||
+	    output->cursor_plane->base.y != y) {
 		if (drmModeMoveCursor(b->drm.fd, output->crtc_id, x, y)) {
 			weston_log("failed to move cursor: %m\n");
 			b->cursors_are_broken = 1;
 		}
 
-		output->cursor_plane.x = x;
-		output->cursor_plane.y = y;
+		output->cursor_plane->base.x = x;
+		output->cursor_plane->base.y = y;
 	}
 }
 
@@ -1674,7 +1680,8 @@ drm_assign_planes(struct weston_output *output_base)
 					      &ev->transform.boundingbox);
 
 		if (next_plane == primary ||
-		    next_plane == &output->cursor_plane) {
+		    (output->cursor_plane &&
+		     next_plane == &output->cursor_plane->base)) {
 			/* cursor plane involves a copy */
 			ev->psf_flags = 0;
 		} else {
@@ -1731,7 +1738,6 @@ drm_output_destroy(struct weston_output *output_base)
 	}
 
 	weston_plane_release(&output->fb_plane);
-	weston_plane_release(&output->cursor_plane);
 
 	weston_output_destroy(&output->base);
 
@@ -2664,6 +2670,59 @@ connector_get_current_mode(drmModeConnector *connector, int drm_fd,
 	return 0;
 }
 
+static void
+drm_output_init_cursor(struct drm_output *output)
+{
+	struct drm_backend *b =
+		(struct drm_backend *)output->base.compositor->backend;
+	struct drm_plane *plane;
+
+	if (b->universal_planes) {
+		wl_list_for_each(plane, &b->plane_list, link) {
+			if (plane->type != WDRM_PLANE_TYPE_CURSOR)
+				continue;
+			if (plane->output)
+				continue;
+			if (!drm_plane_crtc_supported(output,
+						      plane->possible_crtcs))
+				continue;
+
+			plane->output = output;
+			output->cursor_plane = plane;
+			break;
+		}
+	}
+	else {
+		/* XXX: Gross open-coding ... ? */
+		plane = zalloc(sizeof(*plane) + sizeof(uint32_t));
+		if (!plane) {
+			weston_log("%s: out of memory\n", __func__);
+			return;
+		}
+
+		weston_plane_init(&plane->base, b->compositor, 0, 0);
+		wl_list_insert(&b->plane_list, &plane->link);
+
+		plane->plane_id = 0;
+		plane->possible_crtcs = 0;
+		plane->output = output;
+		plane->current = NULL;
+		plane->next = NULL;
+		plane->backend = b;
+		plane->count_formats = 1;
+		plane->formats[0] = GBM_FORMAT_ARGB8888;
+		plane->type = WDRM_PLANE_TYPE_CURSOR;
+
+		output->cursor_plane = plane;
+	}
+
+	if (!output->cursor_plane)
+		return;
+
+	weston_compositor_stack_plane(b->compositor, &output->cursor_plane->base,
+				      NULL);
+}
+
 /**
  * Create and configure a Weston output structure
  *
@@ -2785,6 +2844,8 @@ create_output_for_connector(struct drm_backend *b,
 			   connector->mmWidth, connector->mmHeight,
 			   transform, scale);
 
+	drm_output_init_cursor(output);
+	
 	if (b->use_pixman) {
 		if (drm_output_init_pixman(output, b) < 0) {
 			weston_log("Failed to init output pixman state\n");
@@ -2822,10 +2883,7 @@ create_output_for_connector(struct drm_backend *b,
 	output->base.gamma_size = output->original_crtc->gamma_size;
 	output->base.set_gamma = drm_output_set_gamma;
 
-	weston_plane_init(&output->cursor_plane, b->compositor, 0, 0);
 	weston_plane_init(&output->fb_plane, b->compositor, 0, 0);
-
-	weston_compositor_stack_plane(b->compositor, &output->cursor_plane, NULL);
 	weston_compositor_stack_plane(b->compositor, &output->fb_plane,
 				      &b->compositor->primary_plane);
 
@@ -3258,7 +3316,9 @@ session_notify(struct wl_listener *listener, void *data)
 
 		wl_list_for_each(output, &compositor->output_list, base.link) {
 			output->base.repaint_needed = 0;
-			drmModeSetCursor(b->drm.fd, output->crtc_id, 0, 0, 0);
+			if (output->cursor_plane)
+				drmModeSetCursor(b->drm.fd, output->crtc_id,
+						 0, 0, 0);
 		}
 
 		output = container_of(compositor->output_list.next,
@@ -3564,6 +3624,7 @@ drm_backend_create(struct weston_compositor *compositor,
 		goto err_base;
 
 	b->use_pixman = param->use_pixman;
+	compositor->backend = &b->base;
 
 	/* Check if we run drm-backend using weston-launch */
 	compositor->launcher = weston_launcher_connect(compositor, param->tty,
@@ -3679,7 +3740,6 @@ drm_backend_create(struct weston_compositor *compositor,
 				   "support failed.\n");
 	}
 
-	compositor->backend = &b->base;
 
 	return b;
 
-- 
2.10.1

